File: backend/govinda-masterdata/src/main/java/net/voytrex/govinda/masterdata/api/PersonController.java
/*
 * Govinda ERP - Person REST Controller
 * Copyright 2026 Voytrex
 * SPDX-License-Identifier: Apache-2.0
 */

package net.voytrex.govinda.masterdata.api;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import java.net.URI;
import java.time.LocalDate;
import java.util.List;
import java.util.Locale;
import java.util.UUID;
import net.voytrex.govinda.common.api.PageResponse;
import net.voytrex.govinda.masterdata.application.ChangeMaritalStatusCommand;
import net.voytrex.govinda.masterdata.application.ChangeNameCommand;
import net.voytrex.govinda.masterdata.application.CreatePersonCommand;
import net.voytrex.govinda.masterdata.application.PersonSearchQuery;
import net.voytrex.govinda.masterdata.application.PersonService;
import net.voytrex.govinda.masterdata.application.UpdatePersonCommand;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/v1/masterdata/persons")
@Tag(name = "Persons", description = "Person management (Versicherte)")
public class PersonController {
    private static final String BEARER_AUTH = "bearerAuth";
    private static final String TENANT_HEADER = "X-Tenant-Id";
    private static final String READ_AUTHORITY = "hasAuthority('person:read')";
    private static final String WRITE_AUTHORITY = "hasAuthority('person:write')";

    private final PersonService personService;

    public PersonController(PersonService personService) {
        this.personService = personService;
    }

    @GetMapping
    @Operation(
        summary = "List persons",
        description = "Returns paginated list of persons",
        security = @SecurityRequirement(name = BEARER_AUTH)
    )
    @PreAuthorize(READ_AUTHORITY)
    public PageResponse<PersonResponse> listPersons(
        @RequestHeader(TENANT_HEADER) UUID tenantId,
        @RequestParam(defaultValue = "0") int page,
        @RequestParam(defaultValue = "20") int size,
        @RequestParam(defaultValue = "lastName") String sortBy,
        @RequestParam(defaultValue = "ASC") String sortDir
    ) {
        var pageable = PageRequest.of(
            page,
            Math.min(size, 100),
            Sort.by(Sort.Direction.valueOf(sortDir.toUpperCase(Locale.ROOT)), sortBy)
        );
        var result = personService.listPersons(tenantId, pageable);
        return new PageResponse<>(
            result.getContent().stream().map(PersonMapper::toResponse).toList(),
            result.getNumber(),
            result.getSize(),
            result.getTotalElements(),
            result.getTotalPages(),
            result.isFirst(),
            result.isLast()
        );
    }

    @GetMapping("/search")
    @Operation(
        summary = "Search persons",
        description = "Search persons by various criteria",
        security = @SecurityRequirement(name = BEARER_AUTH)
    )
    @PreAuthorize(READ_AUTHORITY)
    // CHECKSTYLE:OFF: ParameterNumber - Search method requires multiple optional parameters
    public PageResponse<PersonResponse> searchPersons(
        @RequestHeader(TENANT_HEADER) UUID tenantId,
        @RequestParam(required = false) String lastName,
        @RequestParam(required = false) String firstName,
        @RequestParam(required = false) String ahvNr,
        @RequestParam(required = false) LocalDate dateOfBirth,
        @RequestParam(required = false) String postalCode,
        @RequestParam(defaultValue = "0") int page,
        @RequestParam(defaultValue = "20") int size
    ) {
        // CHECKSTYLE:ON: ParameterNumber
        var query = new PersonSearchQuery(
            tenantId,
            lastName,
            firstName,
            ahvNr,
            dateOfBirth,
            postalCode
        );
        var pageable = PageRequest.of(page, Math.min(size, 100));
        var result = personService.searchPersons(query, pageable);
        return new PageResponse<>(
            result.getContent().stream().map(PersonMapper::toResponse).toList(),
            result.getNumber(),
            result.getSize(),
            result.getTotalElements(),
            result.getTotalPages(),
            result.isFirst(),
            result.isLast()
        );
    }

    @GetMapping("/{id}")
    @Operation(
        summary = "Get person",
        description = "Returns a person by ID",
        security = @SecurityRequirement(name = BEARER_AUTH)
    )
    @PreAuthorize(READ_AUTHORITY)
    public PersonResponse getPerson(
        @RequestHeader(TENANT_HEADER) UUID tenantId,
        @PathVariable UUID id
    ) {
        return PersonMapper.toResponse(personService.getPerson(id, tenantId));
    }

    @PostMapping
    @Operation(
        summary = "Create person",
        description = "Creates a new person",
        security = @SecurityRequirement(name = BEARER_AUTH)
    )
    @PreAuthorize(WRITE_AUTHORITY)
    public ResponseEntity<PersonResponse> createPerson(
        @RequestHeader(TENANT_HEADER) UUID tenantId,
        @Valid @RequestBody CreatePersonRequest request
    ) {
        var command = new CreatePersonCommand(
            tenantId,
            request.ahvNr(),
            request.lastName(),
            request.firstName(),
            request.dateOfBirth(),
            request.gender(),
            request.maritalStatus(),
            request.nationality(),
            request.preferredLanguage()
        );
        var person = personService.createPerson(command);
        return ResponseEntity
            .created(URI.create("/api/v1/masterdata/persons/" + person.getId()))
            .body(PersonMapper.toResponse(person));
    }

    @PutMapping("/{id}")
    @Operation(
        summary = "Update person",
        description = "Updates non-historical person data",
        security = @SecurityRequirement(name = BEARER_AUTH)
    )
    @PreAuthorize(WRITE_AUTHORITY)
    public PersonResponse updatePerson(
        @RequestHeader(TENANT_HEADER) UUID tenantId,
        @PathVariable UUID id,
        @Valid @RequestBody UpdatePersonRequest request
    ) {
        var command = new UpdatePersonCommand(
            tenantId,
            id,
            request.nationality(),
            request.preferredLanguage()
        );
        return PersonMapper.toResponse(personService.updatePerson(command));
    }

    @PostMapping("/{id}/name-change")
    @Operation(
        summary = "Change name",
        description = "Changes person's name (creates history)",
        security = @SecurityRequirement(name = BEARER_AUTH)
    )
    @ResponseStatus(HttpStatus.OK)
    @PreAuthorize(WRITE_AUTHORITY)
    public PersonResponse changeName(
        @RequestHeader(TENANT_HEADER) UUID tenantId,
        @RequestHeader("X-User-Id") UUID userId,
        @PathVariable UUID id,
        @Valid @RequestBody ChangeNameRequest request
    ) {
        var command = new ChangeNameCommand(
            tenantId,
            id,
            request.newLastName(),
            request.newFirstName(),
            request.reason(),
            request.effectiveDate(),
            userId
        );
        return PersonMapper.toResponse(personService.changeName(command));
    }

    @PostMapping("/{id}/marital-status-change")
    @Operation(
        summary = "Change marital status",
        description = "Changes person's marital status (creates history)",
        security = @SecurityRequirement(name = BEARER_AUTH)
    )
    @ResponseStatus(HttpStatus.OK)
    @PreAuthorize(WRITE_AUTHORITY)
    public PersonResponse changeMaritalStatus(
        @RequestHeader(TENANT_HEADER) UUID tenantId,
        @RequestHeader("X-User-Id") UUID userId,
        @PathVariable UUID id,
        @Valid @RequestBody ChangeMaritalStatusRequest request
    ) {
        var command = new ChangeMaritalStatusCommand(
            tenantId,
            id,
            request.newStatus(),
            request.reason(),
            request.effectiveDate(),
            userId
        );
        return PersonMapper.toResponse(personService.changeMaritalStatus(command));
    }

    @GetMapping("/{id}/history")
    @Operation(
        summary = "Get person history",
        description = "Returns the change history of a person",
        security = @SecurityRequirement(name = BEARER_AUTH)
    )
    @PreAuthorize(READ_AUTHORITY)
    public List<PersonHistoryResponse> getPersonHistory(
        @RequestHeader(TENANT_HEADER) UUID tenantId,
        @PathVariable UUID id
    ) {
        return personService.getPersonHistory(id, tenantId).stream()
            .map(PersonMapper::toResponse)
            .toList();
    }

    @GetMapping("/{id}/history/at/{date}")
    @Operation(
        summary = "Get person state at date",
        description = "Returns the person's state at a specific date",
        security = @SecurityRequirement(name = BEARER_AUTH)
    )
    @PreAuthorize(READ_AUTHORITY)
    public ResponseEntity<PersonHistoryResponse> getPersonStateAt(
        @RequestHeader(TENANT_HEADER) UUID tenantId,
        @PathVariable UUID id,
        @PathVariable @Parameter(description = "Date in YYYY-MM-DD format") LocalDate date
    ) {
        return personService.getPersonStateAt(id, tenantId, date)
            .map(PersonMapper::toResponse)
            .map(ResponseEntity::ok)
            .orElseGet(() -> ResponseEntity.notFound().build());
    }
}

File: backend/govinda-masterdata/src/main/kotlin/net/voytrex/govinda/masterdata/api/PersonController.kt
/*
 * Govinda ERP - Person REST Controller
 * Copyright 2024 Voytrex
 * SPDX-License-Identifier: Apache-2.0
 */

package net.voytrex.govinda.masterdata.api

import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.Parameter
import io.swagger.v3.oas.annotations.tags.Tag
import jakarta.validation.Valid
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.Past
import net.voytrex.govinda.common.api.PageResponse
import net.voytrex.govinda.common.domain.model.AgeGroup
import net.voytrex.govinda.common.domain.model.Gender
import net.voytrex.govinda.common.domain.model.Language
import net.voytrex.govinda.common.domain.model.MaritalStatus
import net.voytrex.govinda.masterdata.application.*
import net.voytrex.govinda.masterdata.domain.model.Person
import net.voytrex.govinda.masterdata.domain.model.PersonHistoryEntry
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.net.URI
import java.time.LocalDate
import java.util.UUID

@RestController
@RequestMapping("/api/v1/masterdata/persons")
@Tag(name = "Persons", description = "Person management (Versicherte)")
class PersonController(
    private val personService: PersonService
) {

    @GetMapping
    @Operation(summary = "List persons", description = "Returns paginated list of persons")
    fun listPersons(
        @RequestHeader("X-Tenant-Id") tenantId: UUID,
        @RequestParam(defaultValue = "0") page: Int,
        @RequestParam(defaultValue = "20") size: Int,
        @RequestParam(defaultValue = "lastName") sortBy: String,
        @RequestParam(defaultValue = "ASC") sortDir: String
    ): PageResponse<PersonResponse> {
        val pageable = PageRequest.of(
            page,
            size.coerceAtMost(100),
            Sort.by(Sort.Direction.valueOf(sortDir.uppercase()), sortBy)
        )
        val result = personService.listPersons(tenantId, pageable)
        return PageResponse(
            content = result.content.map { it.toResponse() },
            page = result.number,
            size = result.size,
            totalElements = result.totalElements,
            totalPages = result.totalPages,
            first = result.isFirst,
            last = result.isLast
        )
    }

    @GetMapping("/search")
    @Operation(summary = "Search persons", description = "Search persons by various criteria")
    fun searchPersons(
        @RequestHeader("X-Tenant-Id") tenantId: UUID,
        @RequestParam(required = false) lastName: String?,
        @RequestParam(required = false) firstName: String?,
        @RequestParam(required = false) ahvNr: String?,
        @RequestParam(required = false) dateOfBirth: LocalDate?,
        @RequestParam(required = false) postalCode: String?,
        @RequestParam(defaultValue = "0") page: Int,
        @RequestParam(defaultValue = "20") size: Int
    ): PageResponse<PersonResponse> {
        val query = PersonSearchQuery(
            tenantId = tenantId,
            lastName = lastName,
            firstName = firstName,
            ahvNr = ahvNr,
            dateOfBirth = dateOfBirth,
            postalCode = postalCode
        )
        val pageable = PageRequest.of(page, size.coerceAtMost(100))
        val result = personService.searchPersons(query, pageable)
        return PageResponse(
            content = result.content.map { it.toResponse() },
            page = result.number,
            size = result.size,
            totalElements = result.totalElements,
            totalPages = result.totalPages,
            first = result.isFirst,
            last = result.isLast
        )
    }

    @GetMapping("/{id}")
    @Operation(summary = "Get person", description = "Returns a person by ID")
    fun getPerson(
        @RequestHeader("X-Tenant-Id") tenantId: UUID,
        @PathVariable id: UUID
    ): PersonResponse {
        return personService.getPerson(id, tenantId).toResponse()
    }

    @PostMapping
    @Operation(summary = "Create person", description = "Creates a new person")
    fun createPerson(
        @RequestHeader("X-Tenant-Id") tenantId: UUID,
        @Valid @RequestBody request: CreatePersonRequest
    ): ResponseEntity<PersonResponse> {
        val command = CreatePersonCommand(
            tenantId = tenantId,
            ahvNr = request.ahvNr,
            lastName = request.lastName,
            firstName = request.firstName,
            dateOfBirth = request.dateOfBirth,
            gender = request.gender,
            maritalStatus = request.maritalStatus,
            nationality = request.nationality,
            preferredLanguage = request.preferredLanguage
        )
        val person = personService.createPerson(command)
        return ResponseEntity
            .created(URI("/api/v1/masterdata/persons/${person.id}"))
            .body(person.toResponse())
    }

    @PutMapping("/{id}")
    @Operation(summary = "Update person", description = "Updates non-historical person data")
    fun updatePerson(
        @RequestHeader("X-Tenant-Id") tenantId: UUID,
        @PathVariable id: UUID,
        @Valid @RequestBody request: UpdatePersonRequest
    ): PersonResponse {
        val command = UpdatePersonCommand(
            tenantId = tenantId,
            personId = id,
            nationality = request.nationality,
            preferredLanguage = request.preferredLanguage
        )
        return personService.updatePerson(command).toResponse()
    }

    @PostMapping("/{id}/name-change")
    @Operation(summary = "Change name", description = "Changes person's name (creates history)")
    @ResponseStatus(HttpStatus.OK)
    fun changeName(
        @RequestHeader("X-Tenant-Id") tenantId: UUID,
        @RequestHeader("X-User-Id") userId: UUID,
        @PathVariable id: UUID,
        @Valid @RequestBody request: ChangeNameRequest
    ): PersonResponse {
        val command = ChangeNameCommand(
            tenantId = tenantId,
            personId = id,
            newLastName = request.newLastName,
            newFirstName = request.newFirstName,
            reason = request.reason,
            effectiveDate = request.effectiveDate,
            changedBy = userId
        )
        return personService.changeName(command).toResponse()
    }

    @PostMapping("/{id}/marital-status-change")
    @Operation(summary = "Change marital status", description = "Changes person's marital status (creates history)")
    @ResponseStatus(HttpStatus.OK)
    fun changeMaritalStatus(
        @RequestHeader("X-Tenant-Id") tenantId: UUID,
        @RequestHeader("X-User-Id") userId: UUID,
        @PathVariable id: UUID,
        @Valid @RequestBody request: ChangeMaritalStatusRequest
    ): PersonResponse {
        val command = ChangeMaritalStatusCommand(
            tenantId = tenantId,
            personId = id,
            newStatus = request.newStatus,
            reason = request.reason,
            effectiveDate = request.effectiveDate,
            changedBy = userId
        )
        return personService.changeMaritalStatus(command).toResponse()
    }

    @GetMapping("/{id}/history")
    @Operation(summary = "Get person history", description = "Returns the change history of a person")
    fun getPersonHistory(
        @RequestHeader("X-Tenant-Id") tenantId: UUID,
        @PathVariable id: UUID
    ): List<PersonHistoryResponse> {
        return personService.getPersonHistory(id, tenantId).map { it.toResponse() }
    }

    @GetMapping("/{id}/history/at/{date}")
    @Operation(summary = "Get person state at date", description = "Returns the person's state at a specific date")
    fun getPersonStateAt(
        @RequestHeader("X-Tenant-Id") tenantId: UUID,
        @PathVariable id: UUID,
        @PathVariable @Parameter(description = "Date in YYYY-MM-DD format") date: LocalDate
    ): PersonHistoryResponse? {
        return personService.getPersonStateAt(id, tenantId, date)?.toResponse()
    }
}

// ═══════════════════════════════════════════════════════════════
// Request DTOs
// ═══════════════════════════════════════════════════════════════

data class CreatePersonRequest(
    @field:NotBlank(message = "AHV number is required")
    val ahvNr: String,

    @field:NotBlank(message = "Last name is required")
    val lastName: String,

    @field:NotBlank(message = "First name is required")
    val firstName: String,

    @field:Past(message = "Date of birth must be in the past")
    val dateOfBirth: LocalDate,

    val gender: Gender,

    val maritalStatus: MaritalStatus? = null,

    val nationality: String? = "CHE",

    val preferredLanguage: Language? = Language.DE
)

data class UpdatePersonRequest(
    val nationality: String? = null,
    val preferredLanguage: Language? = null
)

data class ChangeNameRequest(
    @field:NotBlank(message = "New last name is required")
    val newLastName: String,

    val newFirstName: String? = null,

    @field:NotBlank(message = "Reason is required")
    val reason: String,

    val effectiveDate: LocalDate
)

data class ChangeMaritalStatusRequest(
    val newStatus: MaritalStatus,

    @field:NotBlank(message = "Reason is required")
    val reason: String,

    val effectiveDate: LocalDate
)

// ═══════════════════════════════════════════════════════════════
// Response DTOs
// ═══════════════════════════════════════════════════════════════

data class PersonResponse(
    val id: UUID,
    val ahvNr: String,
    val lastName: String,
    val firstName: String,
    val fullName: String,
    val dateOfBirth: LocalDate,
    val gender: Gender,
    val age: Int,
    val ageGroup: AgeGroup,
    val maritalStatus: MaritalStatus?,
    val nationality: String?,
    val preferredLanguage: Language,
    val status: String
)

data class PersonHistoryResponse(
    val historyId: UUID,
    val lastName: String,
    val firstName: String,
    val maritalStatus: MaritalStatus?,
    val validFrom: LocalDate,
    val validTo: LocalDate?,
    val mutationType: String,
    val mutationReason: String?,
    val changedBy: UUID,
    val recordedAt: String
)

// ═══════════════════════════════════════════════════════════════
// Mappers
// ═══════════════════════════════════════════════════════════════

fun Person.toResponse() = PersonResponse(
    id = id,
    ahvNr = ahvNr.value,
    lastName = lastName,
    firstName = firstName,
    fullName = fullName(),
    dateOfBirth = dateOfBirth,
    gender = gender,
    age = ageAt(LocalDate.now()),
    ageGroup = ageGroupAt(LocalDate.now()),
    maritalStatus = maritalStatus,
    nationality = nationality,
    preferredLanguage = preferredLanguage,
    status = status.name
)

fun PersonHistoryEntry.toResponse() = PersonHistoryResponse(
    historyId = historyId,
    lastName = lastName,
    firstName = firstName,
    maritalStatus = maritalStatus,
    validFrom = validFrom,
    validTo = validTo,
    mutationType = mutationType.name,
    mutationReason = mutationReason,
    changedBy = changedBy,
    recordedAt = recordedAt.toString()
)
